lib/cn/kvset_split.c

```diff
@@ -0,0 +1,619 @@
+        if (!err)
+            ++tot_keys;
+    }
+
+    if (!err && tot_keys > 0) {
+        err = kbb_finish(kbb, kblks_out);
+        if(ev(!err && kblks_out->n_blks > 1)) {
+            log_info("output kblks %d > 1", kblks_out->n_blks);
+        }
+    }
+
+    wbti_destroy(wbti);
+    kbb_destroy(kbb);
+
+    return err;
+}
+
+/**
+ * kblock_split() - split a given kblock (@kbd) into two kblocks at @split_key
+ *
+ * @kbd:        kblock descriptor describing the source kblock
+ * @split_key:  the key at which the kblock needs to be split.
+ * @kblks_left:  (output) blk list of kblocks containing keys <= split key
+ * @kblks_right: (output) blk list of kblocks containing keys > split key
+ *
+ * NOTE:
+ * If either @kbid_left or @kbid_right is not populated and err == 0, then all
+ * keys from the source kblock got written to either @kbid_left or @kbid_right
+ *
+ * The output mblocks (@kbid_left and @kbid_right) are not committed when this
+ * function returns. It is up to the caller to abort or commit them.
+ */
+static merr_t
+kblock_split(
+    struct kblock_desc   *kbd,
+    const struct key_obj *split_key,
+    struct blk_list      *kblks_left,
+    struct blk_list      *kblks_right)
+{
+    merr_t err;
+
+    INVARIANT(kbd && split_key && kblks_left && kblks_right);
+
+    err = kblock_copy_range(kbd, NULL, split_key, kblks_left);
+    if (!err) {
+        err = kblock_copy_range(kbd, split_key, NULL, kblks_right);
+        if (!err && (kblks_left->n_blks == 0 && kblks_right->n_blks == 0)) {
+            assert(kblks_left->n_blks > 0 || kblks_right->n_blks > 0);
+            err = merr(EBUG);
+        }
+    }
+
+    return err;
+}
+
+/**
+ * Return a split kblock index for the kblocks in a given kvset by comparing the
+ * the min/max keys stored in a kblock header against the split key.
+ *
+ * Return values:
+ *     -1: if all kblocks belongs to the right kvset                         // case 1
+ *      i and *overlap = false: left: [0, i], right [i + 1, nkblks - 1]      // case 2
+ *      i and *overlap = true:  left: [0, ileft], right [iright, nkblks - 1] // case 3
+ */
+static int
+get_kblk_split_index(struct kvset *ks, const struct key_obj *split_kobj, bool *overlap)
+{
+    char split_key[HSE_KVS_KEY_LEN_MAX];
+    uint32_t split_klen;
+    int i;
+
+    *overlap = false;
+
+    key_obj_copy(split_key, sizeof(split_key), &split_klen, split_kobj);
+
+    for (i = 0; i < ks->ks_st.kst_kblks; i++) {
+        struct kvset_kblk *kblk = ks->ks_kblks + i;
+        int rc1;
+
+        rc1 = keycmp(split_key, split_klen, kblk->kb_koff_min, kblk->kb_klen_min);
+        if (rc1 < 0) {
+            return i - 1; /* case 1, 2 */
+        } else {
+            int rc2 = keycmp(split_key, split_klen, kblk->kb_koff_max, kblk->kb_klen_max);
+            if (rc2 <= 0) {
+                *overlap = (rc2 < 0);
+                return i; /* case 2, 3 */
+            }
+        }
+    }
+
+    return ks->ks_st.kst_kblks - 1; /* case 2 */
+}
```

Comment:
Is it necessary to test both min and max keys on each kblock?  Could you instead only test the min key until you're less-than, and then back up and check the max key of the previous block?  The idea being to cut in half the number of compares.  Dunno, just wondering...

Reply:
yep, great optimization!